### ISSUE TITLE
Optional Array Access - Should be Valid

### ISSUE THREAD

**Issue by ricklove** [danielnixon/eslint-plugin-total-functions#35]
The primary way I would like to access an array value is with the `?.` (optional chain operator) to access a value from an array:

```
        const arr: ReadonlyArray<number> = [];
        const foo = arr?.[0]; 
        // Type is number | undefined;
```

This is essentially the same as:

```
        const foo = arr[0] as number | undefined;
```
 

So this should be a valid pattern, not an invalid pattern.

From: https://github.com/danielnixon/eslint-plugin-total-functions/blob/master/src/rules/no-array-subscript.test.ts 

In the invalid patterns:

```
// optional member array access
    {
      filename: "file.ts",
      code: `
        const arr: ReadonlyArray<number> = [];
        const foo = arr?.[0];
      `,
      errors: [
        {
          messageId: "errorStringGeneric",
          type: AST_NODE_TYPES.OptionalMemberExpression,
        },
      ],
    },
```

Am I missing something?

**Comment by danielnixon** [comment]
Hmm, we made that invalid in https://github.com/danielnixon/eslint-plugin-total-functions/issues/25

Maybe @tankers746 could weigh in?

**Comment by danielnixon** [comment]
Okay so the issue is this is mistaken:

```ts
const foo = arr?.[0]; 
// Type is number | undefined;
```

`foo` is actually of type `number`, not `number | undefined`. Try this, it will compile but explode at runtime:

```ts
const arr: ReadonlyArray<number> = [];
const foo = arr?.[0];
foo.toExponential(); // typechecks but explodes
```

**Comment by danielnixon** [comment]
In fact even in this case `foo` is still simply `number`:

```ts
const arr: ReadonlyArray<number> | undefined | null = [];
const foo = arr?.[0];
foo.toExponential(); // typechecks but explodes
```
